File: sentil-java/src/main/java/io/github/sedislab/sentil/MultiMonitor.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.OptionalDouble;

/** Several streaming formulas under one clock. */
public final class MultiMonitor extends NativeResource {
    public MultiMonitor() throws SentilException {
        super(NativeLib.multiMonitorCreate(), NativeLib::multiMonitorDestroy);
    }

    /** Add a formula string under an id. */
    public void add(String id, String formula) throws SentilException {
        NativeLib.multiMonitorAdd(handle(), id, formula);
    }

    /** Add a formula under an id; the formula is borrowed, not consumed. */
    public void add(String id, Formula formula) throws SentilException {
        NativeLib.multiMonitorAddFormula(handle(), id, formula.handle());
    }

    /** Remove the first formula with the id, returning whether one was found. */
    public boolean remove(String id) {
        return NativeLib.multiMonitorRemove(handle(), id);
    }

    /** Clear every monitor's streaming state. */
    public void reset() {
        NativeLib.multiMonitorReset(handle());
    }

    /** The number of formulas. */
    public long size() {
        return NativeLib.multiMonitorLen(handle());
    }

    /** Whether no formula is registered. */
    public boolean isEmpty() {
        return NativeLib.multiMonitorIsEmpty(handle());
    }

    /** The ids in insertion order. */
    public List<String> ids() throws SentilException {
        return Arrays.asList(NativeLib.multiMonitorIds(handle()));
    }

    /**
     * Add a P~p(phi) formula tracked online through a lifted particle ensemble. The
     * formula and registry are borrowed.
     */
    public void addProbabilistic(String id, Formula formula, LiftingRegistry lifting,
            SmcConfig config) throws SentilException {
        NativeLib.multiMonitorAddProbabilistic(handle(), id, formula.handle(), lifting.handle(),
                config.samples, config.confidence, config.seed, config.method.code());
    }

    /** Advance every monitor at this sample, returning the verdict for each id. */
    public Map<String, Robustness> update(double time, Map<String, Double> values)
            throws SentilException {
        NamedSample sample = NamedSample.of(values);
        return NativeLib.multiMonitorUpdate(handle(), time, sample.names, sample.values);
    }

    /** The last satisfaction probability for the formula under {@code id}, or empty. */
    public OptionalDouble probability(String id) {
        double p = NativeLib.multiMonitorProbability(handle(), id);
        return Double.isNaN(p) ? OptionalDouble.empty() : OptionalDouble.of(p);
    }

    /** Each formula's last probability keyed by id, in insertion order. */
    public Map<String, OptionalDouble> probabilities() throws SentilException {
        Map<String, OptionalDouble> out = new LinkedHashMap<>();
        for (String id : ids()) {
            out.put(id, probability(id));
        }
        return out;
    }
}

File: sentil-matlab/+sentil/MultiMonitor.m
classdef MultiMonitor < handle
    %MULTIMONITOR Several named formulas advanced together on one clock.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = MultiMonitor()
            obj.Handle = sentil_mex('multi_monitor_create');
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('multi_monitor_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function add(obj, id, spec)
            %ADD Register a formula under an id, from a string or a borrowed sentil.Formula.
            obj.assertOpen();
            if isa(spec, 'sentil.Formula')
                sentil_mex('multi_monitor_add_formula', obj.Handle, char(id), spec.Handle);
            else
                sentil_mex('multi_monitor_add', obj.Handle, char(id), char(spec));
            end
        end

        function add_probabilistic(obj, id, formula, lifting, config)
            %ADD_PROBABILISTIC Register a P-wrapped formula tracked with a lifted ensemble.
            obj.assertOpen();
            if nargin < 5, config = sentil.SmcConfig; end
            sentil_mex('multi_monitor_add_probabilistic', obj.Handle, char(id), formula.Handle, ...
                lifting.Handle, config.samples, config.confidence, config.seed, ...
                double(int32(config.method)));
        end

        function tf = remove(obj, id)
            %REMOVE Drop a monitor by id.
            obj.assertOpen();
            tf = sentil_mex('multi_monitor_remove', obj.Handle, char(id));
        end

        function reset(obj)
            %RESET Clear every monitor's streaming state.
            obj.assertOpen();
            sentil_mex('multi_monitor_reset', obj.Handle);
        end

        function n = length(obj)
            %LENGTH The number of registered monitors.
            obj.assertOpen();
            n = sentil_mex('multi_monitor_len', obj.Handle);
        end

        function tf = is_empty(obj)
            %IS_EMPTY Whether no monitor is registered.
            obj.assertOpen();
            tf = sentil_mex('multi_monitor_is_empty', obj.Handle);
        end

        function v = ids(obj)
            %IDS The registered ids in insertion order.
            obj.assertOpen();
            v = sentil_mex('multi_monitor_ids', obj.Handle);
        end

        function m = update(obj, time, sample)
            %UPDATE Advance every monitor at one sample, returning verdicts keyed by id.
            obj.assertOpen();
            [names, values] = unpack_sample(sample);
            arr = sentil_mex('multi_monitor_update', obj.Handle, double(time), names, values);
            m = containers.Map('KeyType', 'char', 'ValueType', 'any');
            for i = 1:numel(arr)
                e = arr(i);
                m(e.id) = struct('resolved', e.resolved, 'satisfied', e.satisfied, ...
                    'value', e.value, 'lower', e.lower, 'upper', e.upper);
            end
        end

        function p = probability(obj, id)
            %PROBABILITY The last satisfaction probability for the formula under id.
            obj.assertOpen();
            p = sentil_mex('multi_monitor_probability', obj.Handle, char(id));
        end

        function m = probabilities(obj)
            %PROBABILITIES The last probability of every monitor, keyed by id.
            obj.assertOpen();
            names = obj.ids();
            m = containers.Map('KeyType', 'char', 'ValueType', 'double');
            for i = 1:numel(names)
                m(names{i}) = obj.probability(names{i});
            end
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this monitor has been closed');
            end
        end
    end
end